### ISSUE TITLE
Performance Optimizations for Hacktoberfest

### ISSUE THREAD

**Issue by colinodell** [thephpleague/commonmark#385]
Hacktoberfest is upon us. In order to make Hacktoberfest a positive experience for everybody, this issue describes what kind of PR's are welcomed with open arms.

Due to the advanced logic needed to properly implement the CommonMark spec, this library has always been slower and more memory-intensive than other Markdown libraries for PHP:

```
Benchmark Results, CPU:
	1) Parsedown                    1.44 ms              
	2) cebe/markdown                1.64 ms      -0.20 ms
	3) cebe/markdown extra          1.87 ms      -0.43 ms
	4) cebe/markdown gfm            2.32 ms      -0.88 ms
	5) PHP Markdown                 2.88 ms      -1.44 ms
	6) PHP Markdown Extra           4.14 ms      -2.70 ms
	7) CommonMark                  18.43 ms     -16.99 ms

Benchmark Results, Peak Memory:
	1) PHP Markdown                 2294 kB              
	2) Parsedown                    2437 kB       -143 kB
	3) PHP Markdown Extra           2530 kB       -236 kB
	4) cebe/markdown                2543 kB       -249 kB
	5) cebe/markdown extra          2614 kB       -320 kB
	6) cebe/markdown gfm            2968 kB       -674 kB
	7) CommonMark                   6630 kB      -4336 kB
```

Although these benchmarks aren't that bad and I'm fairly happy with the performance, I constantly wonder whether there are optimizations I haven't considered which can further improve the performance of this library.

**I'd therefore like to welcome the community to help identify any optimizations we might have overlooked!**

## Submission Guidelines

I am specifically looking for optimizations which:

 - Lower CPU time without unreasonable increases in memory
 - Minimize breaks to backward compatibility (BC-breaking changes will be considered if the benefits are significant enough)
 - Don't significantly complicate the code or maintainability
 - Don't negatively hurt performance for common use cases
 - Don't break compatibility with the CommonMark spec
 - **Show measurable results which can be replicated/proven via our current benchmark script, Blackfire, or similar tools**

## Benchmarking Performance

Using a fresh copy of `master`, install all dev dependencies with Composer and then run the following command **with xdebug disabled**:

```sh
./tests/benchmark/benchmark.php -memory
```

That will give you the baseline results.  You can then make changes and re-run the benchmark to see whether performance improved.

-----

Micro-optimizations are welcome provided they meet the guidelines, but please don't submit anything that hasn't been tested and proven to help performance.

Thanks, and Happy Hacktoberfest!

**Comment by magoun** [comment]
I was looking into this last night, but wasn't able to get the benchmark to work with the `-memory` flag. It kept throwing a "failed to close process" error for CommonMark.

The benchmark runs fine for CPU usage. Any thoughts on why the memory benchmark isn't working?

**Comment by colinodell** [comment]
I'm honestly not sure why that might be failing. If you decide to try contributing a PR anyway with just CPU stats that's fine - I can run the memory profiling myself.  (The memory output just makes it easier and quicker for me to spot-check PRs)

**Comment by ricardofiorani** [comment]
@colinodell this is happening for me too, I was getting the "failed to close process" error message. Upon checking the code output of the `proc_close` method I noticed it was returning `127` which is pretty much a "not found" error.
This is due to a problem of the `realpath` method when it's using a dir containing a space.
(https://github.com/thephpleague/commonmark/blob/master/tests/benchmark/benchmark.php#L206)
To solve this, I changed this mentioned line above to escape spaces:
```php
$bin = str_replace(' ','\ ', realpath($config['exec']));
```
